tokio-rustls/src/common/mod.rs

```diff
@@ -258,14 +258,22 @@ where
             // connection with a `CloseNotify` message and no more data will be forthcoming.
             //
             // Rustls yielded more data: advance the buffer, then see if more data is coming.
+            //
+            // We don't need to modify `self.eof` here, because it is only a temporary mark.
             Ok(n) => {
                 buf.advance(n);
                 Poll::Ready(Ok(()))
             }
 
             // Rustls doesn't have more data to yield, but it believes the connection is open.
             Err(ref err) if err.kind() == io::ErrorKind::WouldBlock => {
-                assert!(io_pending, "rustls unexpectedly would block");
+                if !io_pending {
+                    // In theory it will not happen, but if it does, we can try again.
+                    //
+                    // If the rustls state is abnormal, it may cause a cyclic wakeup.
+                    // but tokio's cooperative budget will prevent infinite wakeup.
+                    cx.waker().wake_by_ref();
```

Comment:
Same here: "it will not happen": what exactly? Why will it not happen? There is a lot of context here and gathering it by reading sources across crates is a pain in the ass. Better to explain it more fully here for future reference, IMO.